gnd/src/main/java/com/google/android/gnd/system/MediaUploadManager.java

```diff
@@ -0,0 +1,141 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.android.gnd.system;
+
+import android.graphics.Bitmap;
+import android.net.Uri;
+import android.util.Log;
+import com.google.firebase.storage.FirebaseStorage;
+import com.google.firebase.storage.StorageReference;
+import com.google.firebase.storage.UploadTask;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.Locale;
+import java.util.Objects;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+// TODO: Add column to Observation table for storing uploaded media urls
+// TODO: Synced to remote db as well
+@Singleton
+public class MediaUploadManager {
+
+  private static final String TAG = MediaUploadManager.class.getName();
+  private static final String MEDIA_ROOT_DIR = "uploaded_media";
+  private final SimpleDateFormat dateFormat =
+      new SimpleDateFormat("yyyyMMddHHmmss", Locale.getDefault());
+
+  @Inject
+  MediaUploadManager() {}
+
+  /** Returns a reference to the default Storage bucket. */
+  private FirebaseStorage getStorage() {
+    return FirebaseStorage.getInstance();
+  }
+
+  /** Returns a reference to the root media dir. */
+  private StorageReference getRootMediaDir() {
+    return getStorage().getReference().child(MEDIA_ROOT_DIR);
+  }
+
+  /**
+   * Returns a reference to a object under the root media dir.
+   *
+   * @param fileName Name of the uploaded media
+   */
+  private StorageReference createReference(String fileName) {
+    return getRootMediaDir().child(fileName + '-' + getFilenameSuffix());
+  }
+
+  /** Converts current timestamp to a string to be used a suffix for uploading media. */
+  private String getFilenameSuffix() {
+    return dateFormat.format(new Date());
+  }
+
+  /**
+   * Upload file to Firebase Storage.
+   */
+  public void uploadMediaFromFile(File file, String fileName) {
+    StorageReference reference = createReference(fileName);
+    UploadTask task = reference.putFile(Uri.fromFile(file));
+
+    uploadMediaToFirebaseStorage(task, fileName);
+    fetchDownloadUrl(reference, task);
+  }
+
+  /** Upload bitmap to Firebase Storage. */
+  public void uploadMediaFromBitmap(Bitmap bitmap, String fileName) {
+    ByteArrayOutputStream baos = new ByteArrayOutputStream();
+    bitmap.compress(Bitmap.CompressFormat.JPEG, 100, baos);
+    byte[] data = baos.toByteArray();
+
+    StorageReference reference = createReference(fileName);
+    UploadTask task = reference.putBytes(data);
+
+    uploadMediaToFirebaseStorage(task, fileName);
+    fetchDownloadUrl(reference, task);
+  }
+
+  private void uploadMediaToFirebaseStorage(UploadTask uploadTask, String fileName) {
+    // TODO: Create UploadState enum and use RxJava to broadcast upload state globally.
```

Comment:
FYI for when you get to this, you can use our util RxTask to convert the task to a Single.

Reply:
Note added